File: web/src/lib/messages/loadlevel.ts
import {
  bufferToHeader,
  bufferToString,
  HEADER_SIZE,
  headerToBuffer,
  MessageType,
  type Message,
} from ".";

export default class LoadLevelMessage implements Message {
  id: number;
  sentAt: bigint;
  levelId: string;

  constructor(id: number, sentAt: bigint, levelId: string) {
    this.id = id;
    this.sentAt = sentAt;
    this.levelId = levelId;
  }

  static fromBuffer(buffer: Uint8Array): LoadLevelMessage {
    const header = bufferToHeader(buffer);

    const playerIdOffset = HEADER_SIZE;
    const { value: levelId } = bufferToString(buffer, playerIdOffset);

    return new LoadLevelMessage(header.id, header.sentAt, levelId);
  }

  toBuffer() {
    const encodedHeader = headerToBuffer(this);

    const encoder = new TextEncoder();
    const encodedLevelId = encoder.encode(this.levelId);

    const buffer = new Uint8Array(HEADER_SIZE + 1 + encodedLevelId.byteLength);
    buffer[0] = MessageType.LoadLevel;
    buffer.set(encodedHeader, 1);
    buffer[HEADER_SIZE] = encodedLevelId.byteLength;
    buffer.set(encodedLevelId, HEADER_SIZE + 1);

    return buffer;
  }
}

File: internal/messages/loadlevel.go
package messages

type LoadLevel struct {
	Header
	LevelID string
}

func DecodeLoadLevel(b []byte) (*LoadLevel, error) {
	header, err := DecodeHeader(b[1:HeaderSize])
	if err != nil {
		return nil, err
	}

	if len(b) < HeaderSize+1 {
		return nil, ErrInvalidLoadLevel
	}

	levelID, err := bufferToString(b, HeaderSize)
	if err != nil {
		return nil, err
	}

	return &LoadLevel{
		Header:  header,
		LevelID: levelID,
	}, nil
}

func (l *LoadLevel) Type() MessageType {
	return MessageTypeLoadLevel
}

func (l *LoadLevel) Encode() []byte {
	var b []byte
	b = append(b, byte(MessageTypeLoadLevel))
	b = append(b, l.Header.Encode()...)
	b = append(b, uint8(len(l.LevelID)))
	b = append(b, []byte(l.LevelID)...)

	return b
}
